// @shopify/app-bridge-core/esm/validate/actions/button.js
import { Action, Style } from '../../actions/Button/index.js';
import { matchesObject, composeSchemas, validate, makeOptional, matchesBoolean, matchesString, matchesEnum } from '../type-validate.js';
import { createActionValidator } from '../utils.js';

const buttonSchema = matchesObject({
    disabled: makeOptional(matchesBoolean()),
    label: matchesString(),
    style: makeOptional(matchesEnum(Style)),
    icon: makeOptional(matchesString()),
    loading: makeOptional(matchesBoolean()),
    plain: makeOptional(matchesBoolean()),
});
const buttonSchemaWithId = composeSchemas(matchesObject({
    id: matchesString(),
}), buttonSchema);
function validateProps(props) {
    return validate(props, buttonSchema);
}
function validateAction(action) {
    const validator = createActionValidator(Action, action.type === Action.UPDATE ? buttonSchema : undefined, true, true);
    return validate(action, validator);
}

export { Action, buttonSchema, buttonSchemaWithId, validateAction, validateProps };
