// @shopify/app-bridge-core/esm/validate/utils.js
import { matchesObject, matchesEnum, makeOptional, composeSchemas, matchesString } from './type-validate.js';

function createActionValidator(type, payloadSchema = undefined, payloadRequired = false, idRequired = false) {
    const idSchema = matchesObject({
        id: idRequired ? matchesString() : makeOptional(matchesString()),
    });
    const schema = payloadSchema ? composeSchemas(idSchema, payloadSchema) : idSchema;
    return matchesObject({
        type: matchesEnum(type, {
            message: (_, val) => `The action type \`${val}\` is invalid or unsupported`,
        }),
        payload: payloadRequired ? schema : makeOptional(schema),
    });
}
function isValidRelativePath(path) {
    return typeof path === 'string' && (path === '' || path.startsWith('/'));
}
const relativeUrlSchema = composeSchemas(matchesString(), (value) => isValidRelativePath(value)
    ? undefined
    : [{ error: 'invalid_relative_path', value, message: 'expected string to start with `/`' }]);
const relativePathSchema = matchesObject({
    path: relativeUrlSchema,
});

export { createActionValidator, isValidRelativePath, relativePathSchema, relativeUrlSchema };
