// @shopify/app-bridge-core/validate/actions/button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = exports.validateAction = exports.validateProps = exports.buttonSchemaWithId = exports.buttonSchema = void 0;
var Button_1 = require("../../actions/Button");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return Button_1.Action; } });
var type_validate_1 = require("../type-validate");
var utils_1 = require("../utils");
exports.buttonSchema = (0, type_validate_1.matchesObject)({
    disabled: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    label: (0, type_validate_1.matchesString)(),
    style: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesEnum)(Button_1.Style)),
    icon: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesString)()),
    loading: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    plain: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
});
exports.buttonSchemaWithId = (0, type_validate_1.composeSchemas)((0, type_validate_1.matchesObject)({
    id: (0, type_validate_1.matchesString)(),
}), exports.buttonSchema);
function validateProps(props) {
    return (0, type_validate_1.validate)(props, exports.buttonSchema);
}
exports.validateProps = validateProps;
function validateAction(action) {
    var validator = (0, utils_1.createActionValidator)(Button_1.Action, action.type === Button_1.Action.UPDATE ? exports.buttonSchema : undefined, true, true);
    return (0, type_validate_1.validate)(action, validator);
}
exports.validateAction = validateAction;
