// @shopify/app-bridge-core/validate/actions/contextualSaveBar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = exports.validateAction = exports.validateProps = exports.contextSaveBarSchema = void 0;
var type_validate_1 = require("../type-validate");
var ContextualSaveBar_1 = require("../../actions/ContextualSaveBar");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return ContextualSaveBar_1.Action; } });
var utils_1 = require("../utils");
exports.contextSaveBarSchema = (0, type_validate_1.makeOptional)((0, type_validate_1.matchesObject)({
    fullWidth: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    discardAction: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesObject)({
        disabled: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
        discardConfirmationModal: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    })),
    saveAction: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesObject)({
        disabled: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    })),
    leaveConfirmationDisable: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
}));
function validateProps(props) {
    return (0, type_validate_1.validate)(props, exports.contextSaveBarSchema);
}
exports.validateProps = validateProps;
function validateAction(action) {
    var validator = (0, utils_1.createActionValidator)(ContextualSaveBar_1.Action, exports.contextSaveBarSchema);
    return (0, type_validate_1.validate)(action, validator);
}
exports.validateAction = validateAction;
