// @shopify/app-bridge-core/validate/actions/link.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = exports.validateAction = exports.validateProps = exports.linkActionSchema = exports.linkPropsSchema = void 0;
var AppLink_1 = require("../../actions/Link/AppLink");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return AppLink_1.Action; } });
var type_validate_1 = require("../type-validate");
var utils_1 = require("../utils");
var AllowedRedirectType;
(function (AllowedRedirectType) {
    AllowedRedirectType["APP"] = "APP::NAVIGATION::REDIRECT::APP";
    AllowedRedirectType["LEGACY_APP"] = "APP";
})(AllowedRedirectType || (AllowedRedirectType = {}));
exports.linkPropsSchema = (0, type_validate_1.matchesObject)({
    label: (0, type_validate_1.matchesString)(),
    destination: utils_1.relativeUrlSchema,
    redirectType: (0, type_validate_1.matchesEnum)(AllowedRedirectType),
});
exports.linkActionSchema = (0, type_validate_1.matchesObject)({
    label: (0, type_validate_1.matchesString)(),
    destination: utils_1.relativePathSchema,
    redirectType: (0, type_validate_1.matchesEnum)(AllowedRedirectType),
});
function validateProps(props) {
    return (0, type_validate_1.validate)(props, exports.linkPropsSchema);
}
exports.validateProps = validateProps;
function validateAction(action) {
    return (0, type_validate_1.validate)(action, (0, utils_1.createActionValidator)(AppLink_1.Action, exports.linkActionSchema, true, true));
}
exports.validateAction = validateAction;
