// @shopify/app-bridge-core/validate/actions/toast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = exports.validateProps = exports.validateAction = exports.toastSchema = void 0;
var Toast_1 = require("../../actions/Toast");
Object.defineProperty(exports, "Action", { enumerable: true, get: function () { return Toast_1.Action; } });
var type_validate_1 = require("../type-validate");
var utils_1 = require("../utils");
exports.toastSchema = (0, type_validate_1.matchesObject)({
    message: (0, type_validate_1.matchesString)(),
    duration: (0, type_validate_1.matchesPositiveInteger)(),
    isError: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesBoolean)()),
    action: (0, type_validate_1.makeOptional)((0, type_validate_1.matchesObject)({
        content: (0, type_validate_1.matchesString)(),
    })),
});
function validateAction(action) {
    switch (action.type) {
        case Toast_1.Action.SHOW:
            return (0, type_validate_1.validate)(action, (0, utils_1.createActionValidator)(Toast_1.Action, exports.toastSchema, true));
        default:
            return (0, type_validate_1.validate)(action, (0, utils_1.createActionValidator)(Toast_1.Action));
    }
}
exports.validateAction = validateAction;
function validateProps(props) {
    return (0, type_validate_1.validate)(props, exports.toastSchema);
}
exports.validateProps = validateProps;
