// @shopify/app-bridge-core/esm/actions/Navigation/Redirect/index.js
import { actionWrapper } from '../../helper.js';
import { ActionSet } from '../../ActionSet.js';
import { Group } from '../../types.js';

var Action;
(function (Action) {
    Action["ADMIN_SECTION"] = "APP::NAVIGATION::REDIRECT::ADMIN::SECTION";
    Action["ADMIN_PATH"] = "APP::NAVIGATION::REDIRECT::ADMIN::PATH";
    Action["REMOTE"] = "APP::NAVIGATION::REDIRECT::REMOTE";
    Action["APP"] = "APP::NAVIGATION::REDIRECT::APP";
})(Action || (Action = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["Product"] = "products";
    ResourceType["Collection"] = "collections";
    ResourceType["Order"] = "orders";
    ResourceType["Customer"] = "customers";
    ResourceType["Discount"] = "discounts";
})(ResourceType || (ResourceType = {}));
function isResourcePayload(resource) {
    // tslint:disable-next-line:no-boolean-literal-compare
    return typeof resource.id === 'string';
}
function toAdminPath(payload) {
    return actionWrapper({
        payload,
        group: Group.Navigation,
        type: Action.ADMIN_PATH,
    });
}
function toAdminSection(payload) {
    return actionWrapper({
        payload,
        group: Group.Navigation,
        type: Action.ADMIN_SECTION,
    });
}
function toRemote(payload) {
    return actionWrapper({
        payload,
        group: Group.Navigation,
        type: Action.REMOTE,
    });
}
function toApp(payload) {
    return actionWrapper({
        payload,
        group: Group.Navigation,
        type: Action.APP,
    });
}
function toDestination(action, payload, id) {
    switch (action) {
        case Action.APP: {
            const appPayload = isAppPayload(payload) ? payload : { path: payload };
            return toApp({ id, ...appPayload });
        }
        case Action.ADMIN_PATH: {
            const adminPathPayload = isAdminPathPayload(payload) ? payload : { path: payload };
            return toAdminPath({ id, ...adminPathPayload });
        }
        case Action.ADMIN_SECTION: {
            const adminSectionPayload = isAdminSectionPayload(payload)
                ? payload
                : { section: payload };
            return toAdminSection({ id, ...adminSectionPayload });
        }
        case Action.REMOTE: {
            const remotePayload = isRemotePayload(payload) ? payload : { url: payload };
            return toRemote({ id, ...remotePayload });
        }
    }
}
function getPathWithSearchAndHash({ pathname, search, hash }) {
    return `${pathname}${search || ''}${hash || ''}`;
}
function normalizeUrl(to) {
    if (to instanceof URL) {
        return to.toString();
    }
    if (typeof to === 'string') {
        return to;
    }
    return getRelativePath(to);
}
function getRelativePath(to) {
    if (typeof to === 'string') {
        if (to.startsWith('/')) {
            return to;
        }
        return getPathWithSearchAndHash(new URL(to));
    }
    const search = to.search instanceof URLSearchParams ? to.search.toString() : to.search;
    return getPathWithSearchAndHash({ ...to, search });
}
function isAppPayload(payload) {
    return typeof payload === 'object' && Object.prototype.hasOwnProperty.call(payload, 'path');
}
function isAdminPathPayload(payload) {
    return typeof payload === 'object' && Object.prototype.hasOwnProperty.call(payload, 'path');
}
function isAdminSectionPayload(payload) {
    return (typeof payload === 'object' &&
        typeof payload.section === 'object' &&
        Object.prototype.hasOwnProperty.call(payload.section, 'name'));
}
function isRemotePayload(payload) {
    return typeof payload === 'object' && Object.prototype.hasOwnProperty.call(payload, 'url');
}
function isAdminSection(to) {
    return typeof to === 'object' && typeof to?.name === 'string';
}
function isCreateResourcePayload(resource) {
    // tslint:disable-next-line:no-boolean-literal-compare
    return resource.create === true;
}
function isProductVariantResourcePayload(resource) {
    const castResource = resource;
    // tslint:disable-next-line:no-boolean-literal-compare
    return castResource.id !== undefined && castResource.variant !== undefined;
}
class Redirect extends ActionSet {
    constructor(app) {
        super(app, 'Redirect', Group.Navigation);
    }
    get payload() {
        return { id: this.id };
    }
    dispatch(action, payload) {
        const redirectAction = toDestination(action, payload, this.payload.id);
        this.app.dispatch(redirectAction);
        return this;
    }
}

export { Action, Redirect, ResourceType, getPathWithSearchAndHash, getRelativePath, isAdminPathPayload, isAdminSection, isAdminSectionPayload, isAppPayload, isCreateResourcePayload, isProductVariantResourcePayload, isRemotePayload, isResourcePayload, normalizeUrl, toAdminPath, toAdminSection, toApp, toDestination, toRemote };
